src/d3a/models/market/two_sided.py

```diff
@@ -257,9 +257,9 @@ def accept_bid_offer_pair(self, bid, offer, clearing_rate, trade_bid_info, selec
     def match_recommendations(
             self, recommendations: List[BidOfferMatch.serializable_dict]) -> None:
         """Match a list of bid/offer pairs, create trades and residual offers/bids."""
-        if recommendations is None:
-            return
-        for index, recommended_pair in enumerate(recommendations):
+
+        while recommendations:
```

Comment:
So I understand that you want to digest all recommendations here. But what if e.g. a residual i can not be traded? I guess this will be en endless loop then, right?

Reply:
In the first line I'm popping the object, `recommended_pair = recommendations.pop()`
So we're consuming the list same an a for loop, I suppose no endless loop should happen here

Reply:
OK, I misunderstood the purpose of `_replace_offers_bids_with_residual_in_recommendations_list`. I understood now.